Framework/API/src/FileLoaderRegistry.cpp

```diff
@@ -104,16 +105,23 @@ const boost::shared_ptr<IAlgorithm>
 FileLoaderRegistryImpl::chooseLoader(const std::string &filename) const {
   using Kernel::FileDescriptor;
   using Kernel::NexusDescriptor;
-
+  using Kernel::NexusHDF5Descriptor;
   m_log.debug() << "Trying to find loader for '" << filename << "'\n";
 
   IAlgorithm_sptr bestLoader;
   if (NexusDescriptor::isHDF(filename)) {
```

Comment:
The logic should be something closer to (some functions need to be written):
```C++
if (NexusDescriptor::isHDF(filename)) {
  if (NexusHDF5Descriptor::isReadable(filename)) {
      auto descriptor = NexusHDF5Descriptor(filename);
    bestLoader = searchForLoader<NexusHDF5Descriptor,
                                     IFileLoader<NexusHDF5Descriptor>>(
            filename, m_names[Nexus], m_log);
  }
  if (!bestLoader) {
        bestLoader =
            searchForLoader<NexusDescriptor, IFileLoader<NexusDescriptor>>(
                filename, m_names[Nexus], m_log);
  }
} else {  // non nexus loaders
```

Probably even better if `NexusDescriptor::isHDF` became `NexusDescriptor::isNexus` because that is what it checks. The code inside `NexusHDF5Descriptor::isReadable` should probably just be a copy of the HDF5 portion of [`NexusDescriptor::isHDF`](https://github.com/mantidproject/mantid/blob/master/Framework/Kernel/src/NexusDescriptor.cpp#L52-L86)